internal/lifecycle/readme.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lifecycle
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/GoogleCloudPlatform/serverless-sample-tester/internal/util"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+// The tag that should appear immediately before code blocks in a README to indicate that the enclosed commands
+// are to be used by this program for building and deploying the sample.
+var codeTag = "{sst-run-unix}"
+
+// parseREADME parses a README file with the given name. It reads terminal commands surrounded by one of the codeTags
+// listed above and loads them into a Lifecycle. In the process, it replaces the Cloud Run service name and Container
+// Registry tag with the provided inputs.
+func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] os.Open %s: %w", filename, err)
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	var lifecycle Lifecycle
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.Contains(line, codeTag) {
+			s := scanner.Scan()
+			startCodeBlockLine := scanner.Text()
+
+			c := strings.Contains(startCodeBlockLine, "```")
+			if !s || !c {
+				if err := scanner.Err(); err != nil && !s {
+					return nil, fmt.Errorf("[lifecycle.parseREADME] README bufio.Scanner: %w", err)
+				}
+
+				if !c {
+					return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: expecting start of code block immediately after code tag")
+				} else { // scanner.Scan falied, but no error found. EOF.
+					return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: unexpected EOF; file ended immediately after code tag")
+				}
+			}
+
+			var blockClosed bool
+			for scanner.Scan() {
+				line = scanner.Text()
+				if strings.Contains(line, "```") {
+					blockClosed = true
+					break
+				}
+
+				line = strings.TrimSpace(line)
+
+				// If there is a backslash at the end of the line, this is a multiline command. Keep scanning to get
+				// entire command.
+				for line[len(line)-1] == '\\' {
+					line = line[:len(line)-1]
+
+					scanner.Scan()
```

Comment:
What happens if this Scan is false? (or contains "```"?)

Reply:
Same as here: https://github.com/GoogleCloudPlatform/serverless-sample-tester/pull/2#discussion_r467312902 . Thank you.